near-sdk/src/store/lookup_map/mod.rs

```diff
@@ -0,0 +1,533 @@
+    /// [`BorshSerialize`] and [`ToOwned<Owned = K>`](ToOwned) on the borrowed form *must* match those for
+    /// the key type.
+    pub fn contains_key<Q: ?Sized>(&self, k: &Q) -> bool
+    where
+        K: Borrow<Q>,
+        Q: BorshSerialize + ToOwned<Owned = K> + Ord,
+    {
+        // Check cache before checking storage
+        if self
+            .cache
+            .with_value(&k, |v| v.get().and_then(|s| s.value().as_ref()).is_some())
+            .unwrap_or(false)
+        {
+            return true;
+        }
+        let storage_key = Self::lookup_key(&self.prefix, k);
+        let contains = env::storage_has_key(&storage_key);
+
+        if !contains {
+            // If value not in cache and not in storage, can set a cached `None`
+            let _ = self.cache.get(k.to_owned()).set(CacheEntry::new_cached(None));
+        }
+        contains
+    }
+
+    /// Removes a key from the map, returning the value at the key if the key
+    /// was previously in the map.
+    ///
+    /// The key may be any borrowed form of the map's key type, but
+    /// [`BorshSerialize`] and [`ToOwned<Owned = K>`](ToOwned) on the borrowed form *must* match those for
+    /// the key type.
+    pub fn remove<Q: ?Sized>(&mut self, k: &Q) -> Option<V>
+    where
+        K: Borrow<Q>,
+        Q: BorshSerialize + ToOwned<Owned = K>,
+    {
+        self.get_mut_inner(k).replace(None)
+    }
+
+    /// Gets the given key's corresponding entry in the map for in-place manipulation.
+    /// ```
+    /// use near_sdk::store::LookupMap;
+    ///
+    /// let mut count = LookupMap::new(b"m");
+    ///
+    /// for ch in [7, 2, 4, 7, 4, 1, 7] {
+    ///     let counter = count.entry(ch).or_insert(0);
+    ///     *counter += 1;
+    /// }
+    ///
+    /// assert_eq!(count[&4], 2);
+    /// assert_eq!(count[&7], 3);
+    /// assert_eq!(count[&1], 1);
+    /// assert_eq!(count.get(&8), None);
+    /// ```
+    pub fn entry(&mut self, key: K) -> Entry<K, V>
+    where
+        K: Clone,
+    {
+        let entry = self.get_mut_inner(&key);
+        if entry.value().is_some() {
+            // Value exists in cache and is `Some`
+            Entry::Occupied(OccupiedEntry { key, entry })
+        } else {
+            // Value exists in cache, but is `None`
+            Entry::Vacant(VacantEntry { key, entry })
+        }
+    }
+}
+
+impl<K, V, H> LookupMap<K, V, H>
+where
+    K: BorshSerialize + Ord,
+    V: BorshSerialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    /// Flushes the intermediate values of the map before this is called when the structure is
+    /// [`Drop`]ed. This will write all modified values to storage but keep all cached values
+    /// in memory.
+    pub fn flush(&mut self) {
+        let mut buf = Vec::new();
+        for (k, v) in self.cache.inner().iter_mut() {
+            if let Some(v) = v.get_mut() {
+                if v.is_modified() {
+                    let key = Self::lookup_key(&self.prefix, k);
```

Comment:
thought: looking at this, I don't like it for two reasons:

_First_, we could re-use the buffer inside `lookup_key`

_Second_, I think this means that we hash the data twice. When doing `.insert`, you first hash to check if the data is there, and then do a second `hash` on Drop. 

Does it make sense to always store a hash in a modified entry? Seems that can allow us to remove K: Serialize bound from drop, which might be nice. 

(I don't think we can remove `V: Serilalize` in the same way -- it's important that you can do several in-memory modifications of the value without re-serializing)

Reply:
addr in [`62ca86a` (#487)](https://github.com/near/near-sdk-rs/pull/487/commits/62ca86a10301515c0c3b1eeb74044739eff90645) and [`1f525c8` (#487)](https://github.com/near/near-sdk-rs/pull/487/commits/1f525c8ec0ca9b874853be48e458e3575fef318d). 

Was a little tricky to come up with a solution that fit all cases because it's not always the same. For example, `set` doesn't do any storage reads, so the prefix doesn't need to be calculated there at all, so this solution needed to handle a few permutations for each situation, and it was hard to write this with very simple code. I'll adjust the arbitrary test to try to make sure all cases are tested